pkg/uwb_core: initial support

fjmolinas opened the pull request:
### Contribution description

This PR adds support for Decawave [uwb-core](https://github.com/Decawave/uwb-core), this provide a series of libraries to perform ranging between UWB devices.

The library contains multiple modules and in this initial PR not all are supported, please check the pkg documentation for details. So far TWR is supported, the other modules will be added in subsequent PR's.

For details on the pkg please refer to the pkg documentation.

This PR also includes the `uwb-dw1000`, in the future this could be used to provide a netdev interface for the dw1000 driver.

~~NOTE: 150a798 needs to be fixed, I don't understand the issue yet.~~

### Testing procedure

There is an examples under `examples/twr_aloha`, checking that the example works.

aabadie commented on pkg/uwb-core/include/config/config.h:
```diff
@@ -0,0 +1,370 @@
+/*
```
Do you really need to also import these headers in the RIOT code ?

fjmolinas replied:
Well either I import them or patch the code to comment them out.

aabadie replied:
> patch the code to comment them out.

If you want to skip their content you can just set a CFLAGS to skip the inclusion of their content.

fjmolinas replied:
How?

aabadie replied:
In the package Makefile.include, add `CFLAGS += -DSYS_CONFIG_H` and the content of config.h won't be included. I'm not sure that's what you want btw ;)

fjmolinas replied:
Not exactly, if there is not `config.h` it would still complain about it,  but maybe I can remove `config.h`, its more `mcu.h` that I'm refering to.

fjmolinas commented:
@rdagher might be interested in this one

benpicco commented on cpu/nrf52/periph/spi.c:
```diff
@@ -43,12 +43,14 @@ static mutex_t locks[SPI_NUMOF];
  *          thread until the transfer is done
  */
-static mutex_t busy[SPI_NUMOF];
+// static mutex_t busy[SPI_NUMOF];
```
Why are you removing mutex from the nrf52 SPI driver?
This looks pretty hacky. (not to mention that the radio module can be hooked up to other CPUs as well)

fjmolinas replied:
It is hacky there is a FIXME commit there.

fjmolinas replied:
150a798

fjmolinas replied:
I definitively need to fix this.

fjmolinas replied:
Reverted will remove from history once rebasing.

benpicco commented on pkg/uwb-dw1000/include/syscfg_uwb_dw1000.h:
```diff
@@ -0,0 +1,114 @@
+
+/**
+ * @brief The maximum number of bytes in a single transfer that the
+ *        SPI hardware supports. 255 is safe for nrf52.
```
Uh if SPI peripherals have such limitations, we should amend the `periph/spi` API.
This can lead to some nasty surprises with SPI flash.

fjmolinas replied:
This comes from upstream. might need to be adapted... but there are so many configurations that I started buy simply copying them as well as the defines.

benpicco replied:
Fortunately the SPI driver already [takes care of it](https://github.com/RIOT-OS/RIOT/blob/master/cpu/nrf52/periph/spi.c#L174).

fjmolinas replied:
Still needs to be defined for other platforms though, Should I adapt only the `255 is safe for nrf52` part?

benpicco replied:
I mean it's not wrong that 255 is safe for nrf52 :wink: 
How is this define used?

I see the device can send up to 1kB frames - will the driver split up the transfer?

That should not be necessary if in RIOT the SPI API already takes care if this.

fjmolinas replied:
The dw1000 hardware abstraction layers splits up the input data if larger than, similar to what is already done in the in RIOT spi API, but I'd rather not modify the pkg source code, even if that means some duplicate checks.

benpicco replied:
We could use a higher value though and let the compiler throw it out.

fjmolinas replied:
Hmmm for now it should not matter, it could eventually if we add an asynchronous spi. I'll increase the value and add a message on this

fjmolinas replied:
addressed

fjmolinas commented:
Travis should be happy now, I'll trigger a build to see other issues

fjmolinas commented:
More fixups to make murdock happy, had to blacklist some arch as well as llvm (because of missing stdamoics definitons, should be fixable after #14331 )

fjmolinas commented:
I think I addressed most of murdock complaints, might be missing 2 since the logs where incomplete, I removed the label to not re-trigger a build while no real review has happened yet.

maribu commented:
> More fixups to make murdock happy, had to blacklist some arch as well as llvm (because of missing stdamoics definitons, should be fixable after #14331 )

C11 atomics should work without #14331 - except for AVR due to the old compiler. But that is Ubuntu's fault for not maintaining their packages - upstream AVR support is quite good.

aabadie commented on pkg/uwb-core/include/dpl/dpl_eventq.h:
```diff
@@ -0,0 +1,243 @@
+ *
+ * @file
+ * @brief       uwb-core DPL (Decawave Porting Layer) event queue wrappers
+ *
```
The author is missing in all dpl_xx.h headers.

aabadie commented on pkg/uwb-core/include/dpl/queue.h:
```diff
@@ -0,0 +1,525 @@
+/*
```
Where does this file come from ? Would it possible to not add it to the code base ?

fjmolinas replied:
It comes from the package, I can't add it isolated without creating an include nightmare because of where it is located in that package.

fjmolinas replied:
see 2ce77cd185816fcc8e5529992b22d87df08cc38d

aabadie commented on pkg/uwb-core/patches/0006-fixup-uwb-uwb.c-use-RIOT-specific-uwb_dev_idx_lookup.patch:
```diff
@@ -0,0 +1,25 @@
+ #include <stdlib.h>
+ #endif
+ 
+-#ifndef (RIOT)
```
This is a patch of a patch ? You can drop it :)

fjmolinas replied:
OK

aabadie commented on pkg/uwb-dw1000/Makefile:
```diff
@@ -0,0 +1,28 @@
+  CFLAGS += -Wno-microsoft-anon-tag
+endif
+
+.PHONY: all
```
I don't this line is needed.

aabadie commented on pkg/uwb-dw1000/Makefile.dep:
```diff
@@ -0,0 +1,14 @@
+USEPKG += uwb-core
```
This dependency seems wrong to me. Isn't `uwb-core` that require `uwb-dw1000` ?

fjmolinas replied:
Its not a clear line, `uwb-dw1000` needs the `dpl` part of `uwb-core`. I'll see i I can reflect that dependency better

fjmolinas replied:
For now I'll remove the dependency since `uwb-dw1000` is not yet usable stand alone

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000.c:
```diff
@@ -0,0 +1,175 @@
+#include "utlist.h"
+#include "thread.h"
+
+#ifndef LOG_LEVEL
+#define LOG_LEVEL   LOG_NONE
+#endif
+#include "log.h"
```
I don't understand why you do that all over the place. The log level is something that is set globally using CFLAGS or by setting LOG_LEVEL make variable.
The only thing that should be done with `LOG_` functions is to include `log.h` but not to define a specific default value in each file.

fjmolinas replied:
To enable logs locally.

aabadie replied:
Ok, I understand. It seems to be used already in other places, so I won't insist.

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000_spi.c:
```diff
@@ -0,0 +1,91 @@
+ */
+
+/**
+ * @ingroup     pkg_uwb_dw1000
```
```suggestion
```

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000.c:
```diff
@@ -0,0 +1,175 @@
+ *
+ * @file
+ * @brief       Glue code for running uwb-core for RIOT
+ * *
```
author is missing, same in other files

fjmolinas replied:
Yeah realized its missing from a lot of them

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000.c:
```diff
@@ -0,0 +1,175 @@
+static struct _dw1000_dev_instance_t * dw1000_instance_head;
+
+/* Default instance configuration */
+static const dw1000_dev_instance_t dw1000_instance_config_default = {
```
There are lots of hardcoded default values in this struct. Do you think it would make sense to allow them to be overridable (like `DW1000_ROLE_DEFAULT`) ?

fjmolinas replied:
OK.

fjmolinas replied:
done

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000.c:
```diff
@@ -0,0 +1,175 @@
+{
+    dw1000_dev_instance_t *current = dw1000_instance_head;
+
+    while (current != NULL)
+    {
```
```suggestion
    while (current) {
```

aabadie reviewed (commented):
An initial (big) set of comments. The change set is huge, so I might have missed some implementation and adaptation details.

fjmolinas commented on examples/twr_aloha/Makefile:
```diff
@@ -0,0 +1,60 @@
+DW1000_ROLE ?= 0x00
+CFLAGS += -DDW1000_ROLE_DEFAULT=$(DW1000_ROLE)
+
+# All uwb-core applications need to enable `-fms-extensions`
+CFLAGS += -fms-extensions
+ifneq (,$(filter llvm,$(TOOLCHAIN)))
+  CFLAGS += -Wno-microsoft-anon-tag
+endif
```
I though it could be moved to the package, and it is there as well, but I get the warning when application code makes calls of the api's that have these constructs, I did not find another way to deal with this other than also having the flags in the application. Any ideas on other ways to deal with this?

aabadie replied:
Maybe it could be placed in the package Makefile.include.

fjmolinas replied:
It is already and it was not enough

aabadie replied:
Ok, so let's leave it also at application level then.

fjmolinas commented on examples/twr_aloha/Makefile:
```diff
@@ -0,0 +1,60 @@
+# - UWB_DATA_CODE_SS_TWR_ACK
+# - UWB_DATA_CODE_DS_TWR
+# - UWB_DATA_CODE_DS_TWR_EXT
+UWB_FIXED_TWR_ALGORITHM ?= UWB_DATA_CODE_SS_TWR
```
The current application will send ranging request looping over the different algorithms. You can tell it not to do that and use a fixed one.

As you can see since the following line (with the CFLAG) is commented out, the algorithm is not fixed. I believe this is mentioned in the README. Let me know if there are ways I can improve this.

fjmolinas replied:
Let me know if this one can be resolved

fjmolinas replied:
Can this one be resolved?

fjmolinas commented on examples/twr_aloha/Makefile:
```diff
@@ -0,0 +1,60 @@
+# Uncomment to fix the TWR algoritm
+# CFLAGS += -DUWB_FIXED_TWR_ALGORITHM=$(UWB_FIXED_TWR_ALGORITHM)
+
+# Enable for debugging
+CFLAGS += -DDEBUG_ASSERT_VERBOSE
```
Will remove

fjmolinas commented on examples/twr_aloha/README.md:
```diff
@@ -0,0 +1,58 @@
+
+4. Trying different ranging algorithms
+
+The method for ranging used is selected by modifying the mode variable
+in the uwb_ev_cb function in main.c. By default, it will use one of the
+modes available and setting the `mode` variable accordingly.
```
Note taken!

fjmolinas commented on examples/twr_aloha/README.md:
```diff
@@ -0,0 +1,58 @@
+If multiple modes are enabled it will switch among them, to use only
+one algorithm compile as follows:
+
+    $ UWB_FIXED_TWR_ALGORITHM=UWB_DATA_CODE_SS_TWR make -C examples/twr_aloha/ flash term
```
What about `UWB_TWR_ALGORITHM_ONLY_ONE`?

aabadie replied:
Yes that sounds better

fjmolinas replied:
changed

fjmolinas commented on pkg/uwb-core/Makefile.dep:
```diff
@@ -0,0 +1,28 @@
+endif
+
+# Some stdlib functions used by the pkg are not in avr-gcc
+FEATURES_BLACKLIST += arch_avr8 arch_native
```
I'll add a message

fjmolinas commented on pkg/uwb-core/dpl/config.c:
```diff
@@ -0,0 +1,83 @@
+/*
```
I think I tried to do this and could not, but I'll look at it again to give a proper answer.

fjmolinas replied:
Ok so the reason why I didn't is because the needed function is defined `https://github.com/Decawave/uwb-core/blob/master/porting/dpl_os/src/config.c`, but this file defines a lot of functions I don't really need.

What I could do is add a `porting/riot` directory in the package and hide these files as a patch that I can include from there. Would that be better? I can do the same for the `qeue.h` header

aabadie replied:
> hide these files as a patch that I can include from there

This I don't get.

fjmolinas replied:
Do be able to include them from the package I would need to create new files ot new directories so I can include them withouth including other headers that would mess up everything.

fjmolinas replied:
See 2ce77cd185816fcc8e5529992b22d87df08cc38d

fjmolinas commented on pkg/uwb-core/dpl/dpl_task.c:
```diff
@@ -0,0 +1,67 @@
+#include "dpl/dpl_tasks.h"
+#include "thread.h"
+
+#ifndef LOG_LEVEL
+#define LOG_LEVEL   LOG_INFO
+#endif
+#include "log.h"
```
Debugging and changing the value locally, I can remove if you prefer

fjmolinas commented:
Tackled the easy changes @aabadie, I resolved the change requests I addressed.

fjmolinas commented:
@aabadie all your comments have been addressed

aabadie commented on examples/twr_aloha/Makefile:
```diff
@@ -0,0 +1,57 @@
+CFLAGS += -DDW1000_RX_DIAGNOSTIC=1
+
+# Fix the TWR algorithm:
+# - UWB_DATA_CODE_SS_TWR
+# - UWB_DATA_CODE_SS_TWR_EXT
+# - UWB_DATA_CODE_SS_TWR_ACK
+# - UWB_DATA_CODE_DS_TWR
+# - UWB_DATA_CODE_DS_TWR_EXT
+UWB_TWR_ALGORITHM_ONLY_ONE ?= UWB_DATA_CODE_SS_TWR
```
```suggestion
# - UWB_DATA_CODE_TWR_SS
# - UWB_DATA_CODE_TWR_SS_EXT
# - UWB_DATA_CODE_TWR_SS_ACK
# - UWB_DATA_CODE_TWR_DS
# - UWB_DATA_CODE_TWR_DS_EXT
UWB_TWR_ALGORITHM_ONLY_ONE ?= UWB_DATA_CODE_TWR_SS
```
So this will reflect the module names ?

fjmolinas replied:
These definitions comes from the pkg itself https://github.com/Decawave/uwb-core/blob/591b800d81fe435dc103bd00d23eebc9d8e23ea9/hw/drivers/uwb/include/uwb/uwb_ftypes.h#L72

aabadie commented on examples/twr_aloha/control.c:
```diff
@@ -0,0 +1,268 @@
+    }
+}
+
+void init_ranging(void *arg)
```
Since is not used and this function is declared in the application, you could get rid of it:
```suggestion
void init_ranging(void)
```

aabadie commented on examples/twr_aloha/control.c:
```diff
@@ -0,0 +1,268 @@
+    uwb_mac_append_interface(udev, &_uwb_mac_cbs);
+
+    uint32_t utime = xtimer_now_usec();
+    (void)utime;
```
```suggestion
```

aabadie replied:
Ah wait, utime is not used if log level is > debug. What about:
```suggestion
    if (LOG_LEVEL > LOG_DEBUG) {
        (void)utime;
    }
```
? (untested)

aabadie commented on examples/twr_aloha/control.c:
```diff
@@ -0,0 +1,268 @@
+              (uint16_t)ceilf(uwb_dwt_usecs_to_usecs(rng->config.
+                                                     tx_holdoff_delay)));
+
+#if MYNEWT_VAL(TWR_SS_ACK_ENABLED)
```
Is it possible to use RIOT macros (and not `MYNEWT_VAL`) ?

aabadie commented on pkg/uwb-core/Makefile.dep:
```diff
@@ -0,0 +1,36 @@
+# Some stdlib functions used by the pkg are not in avr-gcc
+FEATURES_BLACKLIST += arch_avr8
+# uwb-core has specific compilation sources when compiling kernel
+# libraries these introduce aditional compilation issues that have not
```
```suggestion
# libraries these introduce additional compilation issues that have not
```

aabadie commented:
I tested this PR on IoT-LAB (5 new dwm1001 were deployed recently) and the measured precision is quite good. I tried the different TWR algorithms and they are all working (the expected distance between dwm1001-1 and dwm1001-5 is 4.72m).

- `UWB_DATA_CODE_SS_TWR`

<details>

```
{"utime": 218,"msg": "dw1000_dev_init"}
dpl: starting thread uwb_irq
{"utime": 17071,"msg": "rng_pkg_init"}
{"utime": 20504,"msg": "twr_ss_pkg_init"}
{"utime": 24174,"msg": "twr_ss_ack_pkg_init"}
{"utime": 28193,"msg": "twr_ss_ext_pkg_init"}
{"utime": 32213,"msg": "twr_ds_pkg_init"}
{"utime": 35887,"msg": "twr_ds_ext_pkg_init"}
main(): This is RIOT! (Version: 2021.01-devel-78-gefb18d-review_uwb)
pkg uwb-dw1000 + uwb-core test application
{"utime": 49643,"exec": "/work/riot/RIOT/examples/twr_aloha/control.c"}
{"device_id"="deca0130","panid="DECA","addr"="989C","part_id"="cae2989c","lot_id"="402c188"}
{"utime": 49643,"msg": "frame_duration = 201 usec"}
{"utime": 49643,"msg": "SHR_duration = 139 usec"}
{"utime": 49643,"msg": "holdoff = 821 usec"}
Node role: TAG 
> range start
range start
Start ranging
> {"utime": 40799738,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.570179],"rssi": [-88.290184],"los": [0.000000]}
{"utime": 40930906,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.495612],"rssi": [-87.950569],"los": [0.000000]}
{"utime": 41015452,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.529869],"rssi": [-88.290184],"los": [0.000000]}
{"utime": 41059475,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.574209],"rssi": [-88.158744],"los": [0.000000]}
{"utime": 41104284,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.557844],"rssi": [-88.137847],"los": [0.000000]}
{"utime": 41146974,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.577741],"rssi": [-87.850112],"los": [0.000000]}
{"utime": 41231520,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.573623],"rssi": [-88.137847],"los": [0.000000]}
{"utime": 41275543,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.546100],"rssi": [-88.117057],"los": [0.000000]}
```

</details>

- `UWB_DATA_CODE_SS_TWR_EXT`

<details>

```
{"utime": 217,"msg": "dw1000_dev_init"}
dpl: starting thread uwb_irq
{"utime": 17070,"msg": "rng_pkg_init"}
{"utime": 20503,"msg": "twr_ss_pkg_init"}
{"utime": 24173,"msg": "twr_ss_ack_pkg_init"}
{"utime": 28192,"msg": "twr_ss_ext_pkg_init"}
{"utime": 32212,"msg": "twr_ds_pkg_init"}
{"utime": 35887,"msg": "twr_ds_ext_pkg_init"}
main(): This is RIOT! (Version: 2021.01-devel-78-gefb18d-review_uwb)
pkg uwb-dw1000 + uwb-core test application
{"utime": 49642,"exec": "/work/riot/RIOT/examples/twr_aloha/control.c"}
{"device_id"="deca0130","panid="DECA","addr"="989C","part_id"="cae2989c","lot_id"="402c188"}
{"utime": 49642,"msg": "frame_duration = 201 usec"}
{"utime": 49642,"msg": "SHR_duration = 139 usec"}
{"utime": 49642,"msg": "holdoff = 821 usec"}
Node role: TAG 
> range start
range start
Start ranging
> {"utime": 3835609,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.575361],"rssi": [-88.333572],"los": [0.000000]}
{"utime": 3966777,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.606994],"rssi": [-88.493057],"los": [0.000000]}
{"utime": 4051323,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.637519],"rssi": [-88.515731],"los": [0.000000]}
{"utime": 4095346,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.604087],"rssi": [-88.311828],"los": [0.000000]}
{"utime": 4140155,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.566468],"rssi": [-88.355438],"los": [0.000000]}
{"utime": 4182838,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.571336],"rssi": [-88.012718],"los": [0.000000]}
{"utime": 4267392,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.571664],"rssi": [-88.245285],"los": [0.000000]}
{"utime": 4311415,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.577123],"rssi": [-88.033020],"los": [0.000000]}
{"utime": 4356224,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.597087],"rssi": [-87.990676],"los": [0.000000]}
{"utime": 4398907,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.568252],"rssi": [-87.990676],"los": [0.000000]}
```

</details>

- `UWB_DATA_CODE_SS_TWR_ACK`

<details>

```
{"utime": 217,"msg": "dw1000_dev_init"}
dpl: starting thread uwb_irq
{"utime": 17079,"msg": "rng_pkg_init"}
{"utime": 20512,"msg": "twr_ss_pkg_init"}
{"utime": 24182,"msg": "twr_ss_ack_pkg_init"}
{"utime": 28201,"msg": "twr_ss_ext_pkg_init"}
{"utime": 32221,"msg": "twr_ds_pkg_init"}
{"utime": 35895,"msg": "twr_ds_ext_pkg_init"}
main(): This is RIOT! (Version: 2021.01-devel-78-gefb18d-review_uwb)
pkg uwb-dw1000 + uwb-core test application
{"utime": 49651,"exec": "/work/riot/RIOT/examples/twr_aloha/control.c"}
{"device_id"="deca0130","panid="DECA","addr"="989C","part_id"="cae2989c","lot_id"="402c188"}
{"utime": 49651,"msg": "frame_duration = 201 usec"}
{"utime": 49651,"msg": "SHR_duration = 139 usec"}
{"utime": 49651,"msg": "holdoff = 821 usec"}
Node role: TAG 
> range start
range start
Start ranging
> {"utime": 3505358,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.561713],"rssi": [-88.538536],"los": [0.000000]}
{"utime": 3636526,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.602430],"rssi": [-88.202758],"los": [0.000000]}
{"utime": 3721072,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.580643],"rssi": [-88.470497],"los": [0.000000]}
{"utime": 3765095,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.583068],"rssi": [-88.515731],"los": [0.000000]}
{"utime": 3809904,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.591281],"rssi": [-88.223968],"los": [0.000000]}
{"utime": 3852587,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.566892],"rssi": [-88.446014],"los": [0.000000]}
{"utime": 3937132,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.563466],"rssi": [-88.470497],"los": [0.000000]}
{"utime": 3981155,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.576010],"rssi": [-88.290184],"los": [0.000000]}
{"utime": 4025972,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.568954],"rssi": [-88.223968],"los": [0.000000]}
```

</details>

- `UWB_DATA_CODE_DS_TWR`

<details>

```
{"utime": 218,"msg": "dw1000_dev_init"}
dpl: starting thread uwb_irq
{"utime": 17079,"msg": "rng_pkg_init"}
{"utime": 20512,"msg": "twr_ss_pkg_init"}
{"utime": 24182,"msg": "twr_ss_ack_pkg_init"}
{"utime": 28201,"msg": "twr_ss_ext_pkg_init"}
{"utime": 32221,"msg": "twr_ds_pkg_init"}
{"utime": 35896,"msg": "twr_ds_ext_pkg_init"}
main(): This is RIOT! (Version: 2021.01-devel-78-gefb18d-review_uwb)
pkg uwb-dw1000 + uwb-core test application
{"utime": 49651,"exec": "/work/riot/RIOT/examples/twr_aloha/control.c"}
{"device_id"="deca0130","panid="DECA","addr"="989C","part_id"="cae2989c","lot_id"="402c188"}
{"utime": 49651,"msg": "frame_duration = 201 usec"}
{"utime": 49651,"msg": "SHR_duration = 139 usec"}
{"utime": 49651,"msg": "holdoff = 821 usec"}
Node role: TAG 
> range start
range start
Start ranging
> {"utime": 4284144,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.552608],"rssi": [-88.355438],"los": [0.000000]}
{"utime": 4415312,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.579110],"rssi": [-88.446014],"los": [0.000000]}
{"utime": 4499858,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.542129],"rssi": [-88.401496],"los": [0.000000]}
{"utime": 4543873,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.596262],"rssi": [-88.268646],"los": [0.000000]}
{"utime": 4588690,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.596225],"rssi": [-88.202758],"los": [0.000000]}
{"utime": 4631381,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.614417],"rssi": [-88.538536],"los": [0.000000]}
{"utime": 4715927,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.610880],"rssi": [-88.682189],"los": [0.000000]}
{"utime": 4759950,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.624316],"rssi": [-88.633064],"los": [0.000000]}
{"utime": 4804759,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.594917],"rssi": [-88.538536],"los": [0.000000]}
{"utime": 4847450,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.558114],"rssi": [-88.379409],"los": [0.000000]}
```

</details>

- `UWB_DATA_CODE_DS_TWR_EXT`

<details>

```
{"utime": 218,"msg": "dw1000_dev_init"}
dpl: starting thread uwb_irq
{"utime": 17071,"msg": "rng_pkg_init"}
{"utime": 20503,"msg": "twr_ss_pkg_init"}
{"utime": 24174,"msg": "twr_ss_ack_pkg_init"}
{"utime": 28193,"msg": "twr_ss_ext_pkg_init"}
{"utime": 32212,"msg": "twr_ds_pkg_init"}
{"utime": 35887,"msg": "twr_ds_ext_pkg_init"}
main(): This is RIOT! (Version: 2021.01-devel-78-gefb18d-review_uwb)
pkg uwb-dw1000 + uwb-core test application
{"utime": 49642,"exec": "/work/riot/RIOT/examples/twr_aloha/control.c"}
{"device_id"="deca0130","panid="DECA","addr"="989C","part_id"="cae2989c","lot_id"="402c188"}
{"utime": 49642,"msg": "frame_duration = 201 usec"}
{"utime": 49642,"msg": "SHR_duration = 139 usec"}
{"utime": 49642,"msg": "holdoff = 821 usec"}
Node role: TAG 
> range start
range start
Start ranging
> {"utime": 3270534,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.567840],"rssi": [-88.538536],"los": [0.000000]}
{"utime": 3401702,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.544250],"rssi": [-88.379409],"los": [0.000000]}
{"utime": 3486248,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.617828],"rssi": [-88.493057],"los": [0.000000]}
{"utime": 3530271,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.548103],"rssi": [-88.515731],"los": [0.000000]}
{"utime": 3575073,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.592764],"rssi": [-88.633064],"los": [0.000000]}
{"utime": 3617770,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.587404],"rssi": [-88.379409],"los": [0.000000]}
{"utime": 3702316,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.572762],"rssi": [-88.586593],"los": [0.000000]}
{"utime": 3746339,"c": 291,"uid": 4660,"ouid": 39068,"raz": [4.583347],"rssi": [-88.705879],"los": [0.000000]}
{"utime": 3791148,"c": 296,"uid": 4660,"ouid": 39068,"raz": [4.562242],"rssi": [-88.493057],"los": [0.000000]}
{"utime": 3833838,"c": 274,"uid": 39068,"ouid": 4660,"raz": [4.599436],"rssi": [-88.515731],"los": [0.000000]}
{"utime": 3918384,"c": 278,"uid": 39068,"ouid": 4660,"raz": [4.569953],"rssi": [-88.609771],"los": [0.000000]}
```

</details>

So in the best case we have a 9cm error, in the worst case it's 18cm.

fjmolinas commented:
Changes addressed.

aabadie commented on examples/twr_aloha/README.md:
```diff
@@ -0,0 +1,70 @@
+## Decawave TWR_ALOHA Example
+
+This example allows testing different two-way ranging algorithm between
```
```suggestion
This example aims at testing different two-way ranging algorithms between
```

fjmolinas replied:
Will correct, `algorithms`, I don't agree with the rewording.

aabadie commented on examples/twr_aloha/README.md:
```diff
@@ -0,0 +1,70 @@
+
+4. Trying different ranging algorithms
+
+The method used  for ranging is selected by modifying the `mode` variable
```
```suggestion
The method used for ranging is selected by modifying the `mode` variable
```

aabadie commented on examples/twr_aloha/README.md:
```diff
@@ -0,0 +1,70 @@
+The method used  for ranging is selected by modifying the `mode` variable
+in the uwb_ev_cb function in main.c. By default, it will use one of the
+modes available and will automatically set the `mode` variable accordingly.
+
+If multiple modes are enabled it will switch among them, to use only
+one algorithm compile as follows:
```
This I will rephrase:
```suggestion
If multiple modes are enabled it will switch among them.

To use only one algorithm, compile as follows:
```

I think the first part of sentence refer more to the previous paragraph. The second part, on the opposite, is directly related to the command given after.

aabadie commented on examples/twr_aloha/control.c:
```diff
@@ -0,0 +1,268 @@
+
+#include "shell_commands.h"
+#include "shell.h"
+#include "xtimer.h"
```
xtimer is not used in this file so the include could be removed
```suggestion
```

fjmolinas replied:
There is `uint32_t utime = xtimer_now_usec();`

aabadie replied:
Ah yes, sorry!

aabadie commented on examples/twr_aloha/control.c:
```diff
@@ -0,0 +1,268 @@
+#include "shell.h"
+#include "xtimer.h"
+
+#ifndef LOG_LEVEL
+#define LOG_LEVEL LOG_DEBUG
+#endif
+#include "log.h"
```
In fact, this is an example application, so I would completely replaces the calls to `LOG_DEBUG` functions with plains `puts`/`printf`. And also remove the `LOG_*` defines/include used here.

```suggestion
```

fjmolinas replied:
ok

aabadie commented on pkg/uwb-core/contrib/uwb_core.c:
```diff
@@ -0,0 +1,61 @@
+    return NULL;
+}
+
+event_queue_t * uwb_core_get_eventq(void)
```
```suggestion
event_queue_t *uwb_core_get_eventq(void)
```

aabadie commented on pkg/uwb-core/dpl/dpl_callout.c:
```diff
@@ -0,0 +1,64 @@
+#include <assert.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <string.h>
+#include <signal.h>
+#include <time.h>
```
I don't think all those includes are needed here. But `#include "xtimer.h"` is missing since it's used in this file.

fjmolinas replied:
its included in `dpl_callout.h`

aabadie commented on pkg/uwb-core/dpl/dpl_mutex.c:
```diff
@@ -0,0 +1,62 @@
+#include <errno.h>
+#include <stdio.h>
+
+#include "sema.h"
```
```suggestion
#include "mutex.h"
```

aabadie commented on pkg/uwb-core/dpl/dpl_mutex.c:
```diff
@@ -0,0 +1,62 @@
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
```
errno are not used

aabadie replied:
errno and assert are not used

aabadie commented on pkg/uwb-core/dpl/dpl_sem.c:
```diff
@@ -0,0 +1,63 @@
+ * @}
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <time.h>
+#include <stdio.h>
```
Includes need cleanup I think.

aabadie commented on pkg/uwb-core/include/dpl/dpl_error.h:
```diff
@@ -0,0 +1,59 @@
+#ifndef DPL_DPL_ERROR_H
+#define DPL_DPL_ERROR_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
```
```suggestion
```

aabadie commented on pkg/uwb-core/include/dpl/dpl_mutex.h:
```diff
@@ -0,0 +1,78 @@
+#ifndef DPL_DPL_MUTEX_H
+#define DPL_DPL_MUTEX_H
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
```
```suggestion
```

aabadie commented on pkg/uwb-core/include/dpl/dpl_eventq.h:
```diff
@@ -0,0 +1,245 @@
+#ifndef DPL_DPL_EVENTQ_H
+#define DPL_DPL_EVENTQ_H
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <stddef.h>
```
Are you sure they are needed ?

aabadie commented on pkg/uwb-core/include/dpl/dpl_sem.h:
```diff
@@ -0,0 +1,91 @@
+#ifndef DPL_DPL_SEM_H
+#define DPL_DPL_SEM_H
+
+#include <stdbool.h>
+#include <stddef.h>
```
```suggestion
```

aabadie commented on pkg/uwb-core/include/dpl/dpl_eventq.h:
```diff
@@ -0,0 +1,245 @@
+#ifndef DPL_DPL_EVENTQ_H
+#define DPL_DPL_EVENTQ_H
+
+#include <stdio.h>
+#include <stdbool.h>
```
Are you sure they are needed ?

aabadie commented on pkg/uwb-dw1000/hal/uwb_dw1000.c:
```diff
@@ -0,0 +1,171 @@
+{
+    dw1000_dev_instance_t *current = dw1000_instance_head;
+
+    while (current)
+    {
```
```suggestion
    while (current) {
```

aabadie commented on pkg/uwb-dw1000/include/hal/hal_timer.h:
```diff
@@ -0,0 +1,46 @@
+/**
+ * @brief   The HAL timer structure.
+ */
+struct hal_timer
+{
```
```suggestion
struct hal_timer {
```

aabadie commented on examples/twr_aloha/control.h:
```diff
@@ -0,0 +1,55 @@
+extern "C" {
+#endif
+
+#include <string.h>
+#include <stdio.h>
```
Sorry :eyes: 

```suggestion
```

aabadie commented on pkg/uwb-core/contrib/uwb_core.c:
```diff
@@ -0,0 +1,61 @@
+ * @{
+ *
+ * @file
+ * @brief       uwb-core bootstraping functions
```
```suggestion
 * @brief       uwb-core bootstrapping functions
```

aabadie commented on pkg/uwb-core/contrib/uwb_core_init.c:
```diff
@@ -0,0 +1,79 @@
+ * @{
+ *
+ * @file
+ * @brief       uwb-core bootstraping core
```
```suggestion
 * @brief       uwb-core bootstrapping core
```

Please check if there are others ;)

aabadie commented on pkg/uwb-core/include/dpl/dpl_cputime.h:
```diff
@@ -0,0 +1,160 @@
+extern "C" {
+#endif
+
+#include "xtimer.h"
```
```suggestion
#include <stdint.h>

#include "xtimer.h"
```

aabadie commented on pkg/uwb-core/include/dpl/dpl_callout.h:
```diff
@@ -0,0 +1,85 @@
+#ifndef DPL_DPL_CALLOUT_H
+#define DPL_DPL_CALLOUT_H
+
+#include <stdint.h>
+
```
```suggestion
```

fjmolinas commented:
May I squash @aabadie ?

aabadie commented:
> May I squash @aabadie ?

Yes please!

fjmolinas commented:
Squashed

aabadie commented:
Looks good on Murdock! Just some boards to add to Makefile.ci and missing doxygen documentation

aabadie commented:
Almost there, still one issue :)

aabadie reviewed (approved):
All good on my side. Thanks for the huge work @fjmolinas, that's awesome!

ACK

aabadie commented:
GO!!!
